hie-bios/src/HIE/Bios/GHCApi.hs

```diff
@@ -0,0 +1,284 @@
+
+data Build = CabalPkg | SingleFile deriving Eq
+
+initializeFlagsWithCradle ::
+        (GhcMonad m)
+        => FilePath -- The file we are loading it because of
+        -> Cradle
+        -> m ()
+initializeFlagsWithCradle = initializeFlagsWithCradleWithMessage (Just G.batchMsg)
+
+-- | Initialize the 'DynFlags' relating to the compilation of a single
+-- file or GHC session according to the 'Cradle' and 'Options'
+-- provided.
+initializeFlagsWithCradleWithMessage ::
+        (GhcMonad m)
+        => Maybe G.Messager
+        -> FilePath -- The file we are loading it because of
+        -> Cradle
+        -> m ()
+initializeFlagsWithCradleWithMessage msg fp cradle = do
+      (ex, err, ghcOpts) <- liftIO $ getOptions (cradleOptsProg cradle) fp
+      G.pprTrace "res" (G.text (show (ex, err, ghcOpts, fp))) (return ())
+      case ex of
+        ExitFailure _ -> throwCradleError err
+        _ -> return ()
+      let compOpts = CompilerOptions ghcOpts
+      liftIO $ hPrint stderr ghcOpts
+      initSessionWithMessage msg compOpts
+
+data CradleError = CradleError String deriving (Show)
+
+instance Exception CradleError where
+
+throwCradleError :: GhcMonad m => String -> m ()
+throwCradleError = liftIO . throwIO . CradleError
+
+----------------------------------------------------------------
+cacheDir :: String
+cacheDir = "haskell-ide-engine"
+
+clearInterfaceCache :: FilePath -> IO ()
+clearInterfaceCache fp = do
+  cd <- getCacheDir fp
+  res <- doesPathExist cd
+  when res (removeDirectoryRecursive cd)
+
+getCacheDir :: FilePath -> IO FilePath
+getCacheDir fp = getXdgDirectory XdgCache (cacheDir ++ "/" ++ fp)
+
+initSessionWithMessage :: (GhcMonad m)
+            => Maybe G.Messager
+            -> CompilerOptions
+            -> m ()
+initSessionWithMessage msg CompilerOptions {..} = do
+    df <- G.getSessionDynFlags
+    traceShowM (length ghcOptions)
+
+    let opts_hash = B.unpack $ encode $ H.finalize $ H.updates H.init (map B.pack ghcOptions)
+    fp <- liftIO $ getCacheDir opts_hash
+    -- For now, clear the cache initially rather than persist it across
+    -- sessions
+    liftIO $ clearInterfaceCache opts_hash
+    (df', targets) <- addCmdOpts ghcOptions df
+    void $ G.setSessionDynFlags
+      (disableOptimisation
+      $ setIgnoreInterfacePragmas
+      $ resetPackageDb
+--      $ ignorePackageEnv
+      $ writeInterfaceFiles (Just fp)
+      $ setVerbosity 0
+
+      $ setLinkerOptions df'
+      )
+    G.setLogAction (\_df _wr _s _ss _pp _m -> return ())
+    G.setTargets targets
+    -- Get the module graph using the function `getModuleGraph`
+    mod_graph <- G.depanal [] True
+    void $ G.load' LoadAllTargets msg mod_graph
+
+----------------------------------------------------------------
+
+-- we don't want to generate object code so we compile to bytecode
+-- (HscInterpreted) which implies LinkInMemory
+-- HscInterpreted
+setLinkerOptions :: DynFlags -> DynFlags
+setLinkerOptions df = df {
```

Comment:
See comments around https://github.com/google/haskell-indexer/blob/master/haskell-indexer-backend-ghc/src/Language/Haskell/Indexer/Backend/GhcApiSupport.hs#L140 for edge case quirks expectable with `HscInterpreted`. Is using some Source Plugin (vs GHC API) an alternative here?

Reply:
GHC turns on `HscInterpreted` if it's necessary for a module itself. I havent come across any projects yet which it gets this wrong for.